feat: edit/delete/reply for emoji/sticker/audio

OmarBasem opened the pull request:
#### You can find continuation of this PR at that new [PR](https://github.com/status-im/status-mobile/pull/13808)

Fixes #13246 

### Summary

Add missing actions (edit/delete/reply) to certain types of messages (emoji/sticker/audio)
To implement the needed functionalities there are 2 pull requests to this repo `status-mobile` and [status-go](https://github.com/status-im/status-go/pull/2771).

1. Editing emoji messages: Firstly, I had to allow the emoji content type to be edited from status-go. Secondly, when editing an emoji message, I had to check whether it is still an emoji message or is now a text message, and send the contentType. Also, in status-go I added a "ContentType" field to the "EditMessage" struct and updated the content type when a message is edited. Regarding UI, I had to make sure a white space is added at the end of an emoji message when edited (similar to text messages) to avoid overlapping with the "edited" text and icon.

2. Deleting emoji messages: added delete option on-long-press.

3. Audio message reply: In status-go, I added an AudioLocalURL field to the QuotedMessage struct. When an audio message is replied to, the AudioLocalURL field will be assigned to the QuotedMessage. In status-mobile, inside the quoted-message method, I check if the quoted message contains audio and if so, will render the audio message.

4. Sticker message reply: Initially, I attempted to do it in a similar way to replying to audio, but it didn't work as the sticker url is created using the sticker hash and not the message ID. Currently, the "responseTo" field includes the messageID only. So, what I did is simply get the sticker url from the message itself using the messageID inside the quoted-message method in status-mobile. I added a `HasSticker` field to the QuotedMessage struct in status-go to know if it is a sticker. Additionally, if the sticker is deleted, a message will be shown to the user stating "This message has been deleted!", which I think can also be implemented for replying to other content types.

5. Sticker message delete: added delete option on-long-press.

Sticker reply and delete also work in public chats.

### Review Notes
This PR depends on [that status-go PR](https://github.com/status-im/status-go/pull/2771) for backend functionality: 87db7d953e8fc748baa1fc56a7adffd8fada3055

#### Platforms

- Android
- iOS

status: ready

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 9f7c1af6 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-13719/1/) | 2022-07-25 07:41:06 | ~11 min | `android-e2e` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-13719/1/consoleText) |

Parveshdhull commented:
Hi @OmarBasem, Thank you very much for PR.

Looks like linting and tests are failed.

You can run `make lint` to see the linting errors and fix them manually or run `make lint-fix`.
And you can checkout errors with tests with the help of `make test`.

Parveshdhull commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -86,28 +86,51 @@
+              current-public-key
+              (partial style/quoted-message-author (and outgoing (not pinned)))
+              (and outgoing (not pinned))]]
+            (if (and image
+                     ;; Disabling images for public-chats
+                     (not public?))
+              [fast-image/fast-image
+               {:style  {:width            56
+                         :height           56
+                         :background-color :black
+                         :border-radius    4}
+                :source {:uri image}}]
+
+              (if (and audio
```
Better to use [cond](https://clojuredocs.org/clojure.core/cond), instead of multiple `if`

Parveshdhull commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -86,28 +86,51 @@
+                   [message-status message]]]]
+
+                (if sticker
+                  (if (def repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id))
```
- Please move subscription to letsubs
- Also `(def ...)` returns nil, so `(if (def ...))` will be always false
- @(re-frame/subscribe ...) will return message object and by surrounding this with (), you are using this as a function, wouldn't that will throw error?

OmarBasem replied:
- The line `(if (def repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id))` will be true if the repliedMsg exists (has not been deleted), otherwise will be false. I have refactored it to `letsubs` anyway.
- `(@(re-frame/subscribe [:chats/chat-messages chat-id]) id)` is used to get a particular message (the replied-to message), hence the parentheses are needed. Actually removing the parentheses would throw an error.

OmarBasem commented:
Hi @Parveshdhull, thank you for your comments!

I fixed the linting errors using `make lint-fix`.
I double-checked running the tests using `make test` and all the tests are passing.
I updated to code to use `cond` in the mentioned snippet.
Moved `repliedMsg` subscription to `letsubs`

flexsurfer commented:
hey @OmarBasem thank you for the PR, its little bit difficult to review PR, could you please revert formatting , thanks

OmarBasem commented:
> hey @OmarBasem thank you for the PR, its little bit difficult to review PR, could you please revert formatting , thanks

Hey @flexsurfer, sorry for that, my IDE couldn't format clojurescript properly. I have fixed the formatting.

OmarBasem commented:
Hey @flexsurfer, I am not sure here why the commit cannot be built? It has already been built previously before reformatting. I did not add or remove any code, but now it is not building. I checked lint and tests - they are all passing.

Parveshdhull commented:
> Hey @flexsurfer, I am not sure here why the commit cannot be built? It has already been built previously before reformatting. I did not add or remove any code, but now it is not building. I checked lint and tests - they are all passing.

Hi, you need to rebase your branch with develop
![image](https://user-images.githubusercontent.com/17097240/182409784-d309df75-2ee7-4c0e-87b1-dfafc0f252e0.png)

OmarBasem commented:
Hi @Parveshdhull, I just rebased, and it still cannot be built.

Parveshdhull commented:
> Hi @Parveshdhull, I just rebased, and it still cannot be built.

It is not properly rebased yet, error reproduction steps
```
git clone https://github.com/OmarBasem/status-mobile && cd status-mobile
git remote add status https://github.com/status-im/status-mobile
git fetch status
git rebase status/develop
```

there are still conflicts in `ios/Podfile.lock`

OmarBasem commented:
Hey @Parveshdhull @flexsurfer, I think the PR is ready to be reviewed?

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -87,8 +87,9 @@
-  [_ {:keys [from parsed-text image]} outgoing current-public-key public? pinned]
-  (letsubs [contact-name [:contacts/contact-name-by-identity from]]
+  [_ {:keys [from parsed-text image audio sticker id] :as message} outgoing current-public-key public? pinned chat-id]
+  (letsubs [contact-name [:contacts/contact-name-by-identity from]
+            repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id)] ;; Msg replied to
```
it probably makes sense to stop using `defview`/`letsubs` here

flexsurfer replied:
and better to use get `(get @(re-frame/subscribe [:chats/chat-messages chat-id]) id)`

J-Son89 replied:
I think `repliedMsg`is not following conventions. It should probably be `replied-msg`, and consider using `replied-message` instead as the character saving adds no value imo.

OmarBasem replied:
@flexsurfer but I need to have the replied-message in a variable to use later on in the `if` condition, and when rendering the reply view. How can I do that with `(get @(re-frame/subscribe [:chats/chat-messages chat-id]) id)` ?

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -97,17 +98,33 @@
+       sticker
+       (if repliedMsg
+         [fast-image/fast-image {:style  {:margin 4 :width 56 :height 56}
+           ;; Get the sticker url of the message replied to
+                                 :source {:uri (((repliedMsg :content) :sticker) :url)}}]
+          ;; Let the user know if the message was deleted
+         [react/text {:style (style/quoted-message-text (and outgoing (not pinned)))}
```
Could you please align comments?

OmarBasem replied:
Fixed it

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -467,35 +486,37 @@
-                                       (not @collapsible?)
-                                       (not outgoing)
-                                       (not modal))
+                           #(when
+                             (and (> (.-nativeEvent.layout.height ^js %) max-message-height-px)
+                                  (not @collapsible?)
+                                  (not outgoing)
+                                  (not modal))
```
unnecessary change

OmarBasem replied:
My IDE is doing unnecessary changes to clojurescript formatting. Can you please tell me what IDE are you using?

J-Son89 replied:
are you running `make lint-fix` before commiting?

I'm using vscode but not sure if it's the best option for the tech stack.

OmarBasem replied:
Yes I am running `make lint-fix`. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808)

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -467,35 +486,37 @@
             (when (and @collapsible? (not modal))
               (if @collapsed?
-                (let [color (if pinned colors/pin-background (if mentioned colors/mentioned-background colors/blue-light))]
+                (let [color (if pinned
+                              colors/pin-background
+                              (if mentioned colors/mentioned-background colors/blue-light))]
```
unnecessary change

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -467,35 +486,37 @@
-                                           :start  {:x 0 :y 0} :end {:x 0 :y 0.9}}
-                    [react/view {:height         72 :align-self :center :justify-content :flex-end
-                                 :padding-bottom 10}
+                   {:on-press #(swap! collapsed? not) :style {:position :absolute :bottom 0 :left 0 :right 0 :height 72}}
+                   [react/linear-gradient
+                    {:colors [(str color "00") color]
+                     :start  {:x 0 :y 0}
+                     :end    {:x 0 :y 0.9}}
+                    [react/view {:height 72 :align-self :center :justify-content :flex-end :padding-bottom  10}
```
looks like unnecessary change

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -467,35 +486,37 @@
                        {:color colors/white}]]]]])
-                [react/touchable-highlight {:on-press #(swap! collapsed? not)
-                                            :style    {:align-self :center :margin 5}}
+                [react/touchable-highlight
+                 {:on-press #(swap! collapsed? not) :style {:align-self :center :margin 5}}
```
unnecessary change

rasom commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -520,65 +541,97 @@
+      [{:keys [content current-public-key outgoing edit-enabled public?
+               pinned in-popover? message-pin-enabled content-type edited-at] :as   message}
+       {:keys [on-long-press modal] :as   reaction-picker}]
+               ;; Makes sure to render a text-messsage and not an emoji-message if it has been edited with text
```
align comment

flexsurfer commented on ios/Podfile.lock:
```diff
@@ -635,7 +635,7 @@ SPEC CHECKSUMS:
   FBReactNativeSpec: f2c97f2529dd79c083355182cc158c9f98f4bd6e
   Folly: b73c3869541e86821df3c387eb0af5f65addfab4
-  glog: 36ce0530c6d2c3a5a4326885ef4069564887a1db
+  glog: 64a900d72fb14cd1b2cdf3b717a5555ceb889d49
```
could you please revert this file

OmarBasem replied:
Fixed it

flexsurfer commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -608,27 +661,37 @@
+               [{:on-press #(re-frame/dispatch [:chat.ui/reply-to-message message])
+                 :id       :reply
+                 :label    (i18n/label :t/message-reply)}]
+               (if (and outgoing config/delete-message-enabled?)
```
better to use `when` here

OmarBasem replied:
Fixed it

flexsurfer commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -608,27 +661,37 @@
          [react/view {:style (style/message-view message) :accessibility-label :audio-message}
           [react/view {:style (style/message-view-content)}
-           [message.audio/message-content message] [message-status message]]]]]
+           [react/view {:style (style/style-message-text outgoing)}
```
why extra view ? its better to have less views as possible

OmarBasem replied:
That's not my code. This is showing as a change due to formatting changes by the IDE. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808).

J-Son89 commented on src/status_im/chat/models/input.cljs:
```diff
@@ -252,14 +252,18 @@
                       :js-response true
-                      :on-error #(log/error "failed to edit message " %)
-                      :on-success #(re-frame/dispatch [:sanitize-messages-and-process-response %])}]}
+                      :on-error    #(log/error "failed to edit message " %)
```
any reason you added the whitespace? consider reverting if not.

briansztamfater commented on src/status_im/ui/screens/chat/message/message_old.cljs:
```diff
@@ -97,17 +98,33 @@
+       [fast-image/fast-image
+        {:style  {:width            56
+                  :height           56
+                  :background-color :black
```
Maybe you can use a quo color constant instead of `:black`

OmarBasem replied:
That's not my code. This is showing as a change due to formatting changes by the IDE. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808).

OmarBasem committed:
feat: edit/delte/reply, fixed unnecessary changes

OmarBasem commented:
To fix all the unnecessary formatting changes caused by the IDE, I had to open a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808) @flexsurfer @rasom @J-Son89 @briansztamfater @Parveshdhull
